core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -287,7 +262,81 @@ public void messageReceived(final Request request, final TransportChannel channe
 
         @Override
         public void messageReceived(Request request, TransportChannel channel, Task task) throws Exception {
-            new PrimaryPhase((ReplicationTask) task, request, channel).run();
+            ReplicationTask replicationTask = (ReplicationTask) task;
+            boolean success = false;
+            PrimaryShardReference primaryShardReference = getPrimaryShardReference(request.shardId());
+            try {
+                if (primaryShardReference.isRelocated()) {
+                    setPhase(replicationTask, "primary_delegation");
+                    // delegate primary phase to relocation target
+                    // it is safe to execute primary phase on relocation target as there are no more in-flight operations where primary
+                    // phase is executed on local shard and all subsequent operations are executed on relocation target as primary phase.
+                    final ShardRouting primary = primaryShardReference.routingEntry();
+                    assert primary.relocating() : "indexShard is marked as relocated but routing isn't" + primary;
+                    DiscoveryNode relocatingNode = clusterService.state().nodes().get(primary.relocatingNodeId());
+                    transportService.sendRequest(relocatingNode, transportPrimaryAction, request, transportOptions,
+                        new TransportChannelResponseHandler<Response>(logger, channel, "rerouting indexing to target primary " + primary) {
```

Comment:
We can simply add responseSupplier to the constructor of TransportChannelResponseHandler (same I did in #17752). We can then remove the static methods in that class (one of which should be obsolete anyhow by the change here w.r.t. master).